File: 25-Dynamic-Programming/python/04_0_and_1_knapsack.py
from typing import List

class Solution:
    def knapsack01(self, wt: List[int], val: List[int], n: int, W: int) -> int:
        dp = [0] * (W + 1)

        for i in range(n):
            for w in range(W, wt[i] - 1, -1):
                dp[w] = max(dp[w], val[i] + dp[w - wt[i]])

        return dp[W]

File: 25-Dynamic-Programming/cpp/04-0-and-1-knapsack.cpp
#include <vector>
#include <algorithm>

using namespace std;

class Solution {
public:
    int knapsack01(vector<int>& wt, vector<int>& val, int n, int W) {
        vector<int> dp(W + 1, 0);

        for (int i = 0; i < n; i++) {
            for (int w = W; w >= wt[i]; w--) {
                dp[w] = max(dp[w], val[i] + dp[w - wt[i]]);
            }
        }

        return dp[W];
    }
};
